tfjs-backend-webgpu/perf/tune.js

```diff
@@ -155,70 +130,137 @@ function initPage() {
-    [dimAOuter, dimInner]
+async function timeMatmul(backend, rowNum) {
+  await tf.setBackend(backend);
+  let tensors = [];
+  let tensorsWarmUp = [];
+  let tensorsToDispose = [];
+  let inputs = [];
+
+  // Prepare data
+  const warmA = tf.tensor2d(
+    Array.from({ length: 512 * 512 }, () => Math.floor(Math.random())),
+    [512, 512]
   );
-  const tensorB = tf.tensor2d(
-    Array.from({ length: dimInner * dimBOuter }, () => Math.floor(Math.random())),
-    [dimInner, dimBOuter]
+  const warmB = tf.tensor2d(
+    Array.from({ length: 512 * 512 }, () => Math.floor(Math.random())),
+    [512, 512]
   );
-  document.getElementById('message').innerHTML =
-    `Testing on webgpu, ${STATE.name}, ${STATE.input} ...`;
-
-  // skip the first time
-  await tf.matMul(tensorA, tensorB).data();
+  tensorsWarmUp = { tensorA: warmA, tensorB: warmB };
+  tensorsToDispose.push(warmA);
+  tensorsToDispose.push(warmB);
+  if (rowNum !== undefined) {
+    let input = document.getElementById(`input-${rowNum}`).getAttribute('value');
+    let m = input.split('[')[1].split(',')[0];
+    let k = input.split('[')[2].split(']')[0].split(',')[0];
+    let n = input.split('[')[2].split(']')[0].split(',')[1];
+    inputs = [`${m},${k},${n}`];
+  }
+  else {
+    inputs = INPUTS;
+  }
+  for (let i = 0; i < inputs.length; i++) {
+    let dimAOuter = parseInt(inputs[i].split(',')[0]);
+    let dimInner = parseInt(inputs[i].split(',')[1]);
+    let dimBOuter = parseInt(inputs[i].split(',')[2]);
+    const tensorA = tf.tensor2d(
+      Array.from({ length: dimAOuter * dimInner }, () => Math.floor(Math.random())),
+      [dimAOuter, dimInner]
+    );
+    const tensorB = tf.tensor2d(
+      Array.from({ length: dimInner * dimBOuter }, () => Math.floor(Math.random())),
+      [dimInner, dimBOuter]
+    );
+    tensors.push({ tensorA, tensorB });
+    tensorsToDispose.push(tensorA);
+    tensorsToDispose.push(tensorB);
+  }
 
-  let profile_webgpu, profile_webgl;
+  tf.env().set('CHECK_COMPUTATION_FOR_ERRORS', false);
+  const profile_data = await tf.profile(() => {
+    // Warmup model
+    for (let i = 0; i < inputs.length; i++) {
+      let result = tf.matMul(tensors[i].tensorA, tensors[i].tensorB);
+      result.dispose();
+    }
+    // Warmup gpu, keep gpu frequency at a high level
+    document.getElementById('message').innerHTML =
+      `Warming up testing on ${backend} ...`;
+    for (let i = 0; i < numWarmUp; i++) {
+      let m = tf.matMul(tensorsWarmUp.tensorA, tensorsWarmUp.tensorB);
+      m.dispose();
+    }
+    // Collect result from here
+    for (let i = 0; i < inputs.length; i++) {
+      document.getElementById('message').innerHTML =
+        `Testing on ${backend} ...`;
+      for (let j = 0; j < numAvg; j++) {
+        const result = tf.matMul(tensors[i].tensorA, tensors[i].tensorB);
+        const m = tf.matMul(tensorsWarmUp.tensorA, tensorsWarmUp.tensorB);
+        result.dispose();
+        m.dispose();
+      }
```

Comment:
In `tf.profile`,
1. I think we mainly profile `const result = tf.matMul(tensors[i].tensorA, tensors[i].tensorB)`, would including `warmup model` and `warmup gpu` parts influence the correctness?
2. Why we still need to execute ` tf.matMul(tensorsWarmUp.tensorA, tensorsWarmUp.tensorB)` in the 3rd for-loop, I think the cost time of it may be larger than `tf.matMul(tensors[i].tensorA, tensors[i].tensorB)`, which is the real main workload.
3. How to make sure that all matMul commands are executed one by one? I think they may be executed synchronously here?

Reply:
1. Yes, `warmup model` is for the first run of each matmul shape(costing large shader generate and buffer creating time ... which leads to a significant deviation from the average value) and `warmup gpu` is to make every shape of inputs could be measured in the same frequency.
2. The idea to insert a large warmup tensor between each main workload is also to keep high frequency, the real main workload will pull down the frequency if they execute together, please see below pictures:
**with**: keep frequency at almost 1200M
![image](https://user-images.githubusercontent.com/7088194/195022318-af45c492-68f2-414d-9b30-aa1e81d2a27a.png)
**without**: Visible jitter, break the measurement accurate (only happens on WebGPU)
![image](https://user-images.githubusercontent.com/7088194/195021680-53021254-816c-4640-aeaf-4dd559939396.png)
3. You're right, serial execution would be considered in another PR.